plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java

```diff
@@ -126,8 +136,109 @@ public List<Location> getDescendants(String locationId, Location parentLocation)
+                List<String> locationIds = new ArrayList<>();
+                for (Location location : locations) {
+                    locationIds.add(location.getIdElement().getIdPart());
+                }
+                return getPaginatedLocations(request, locationIds);
+            }
+
+        } else {
+            if (Constants.SyncStrategy.RELATED_ENTITY_LOCATION.equalsIgnoreCase(syncStrategy)
+                    && userRoles.contains(Constants.ROLE_ALL_LOCATIONS)
+                    && !selectedSyncLocations.isEmpty()) {
+                List<LocationHierarchy> locationHierarchies =
+                        getLocationHierarchies(selectedSyncLocations);
+                List<Resource> resourceList =
+                        locationHierarchies.stream()
+                                .map(locationHierarchy -> (Resource) locationHierarchy)
+                                .collect(Collectors.toList());
+                return Utils.createBundle(resourceList);
+            } else {
+                List<LocationHierarchy> locationHierarchies =
+                        practitionerDetailsEndpointHelper
+                                .getPractitionerDetailsByKeycloakId(practitionerId)
+                                .getFhirPractitionerDetails()
+                                .getLocationHierarchyList();
+                List<Resource> resourceList =
+                        locationHierarchies.stream()
+                                .map(locationHierarchy -> (Resource) locationHierarchy)
+                                .collect(Collectors.toList());
+                return Utils.createBundle(resourceList);
+            }
+        }
+    }
+
+    public String getSyncStrategy(DecodedJWT verifiedJwt) {
+        String applicationId = JwtUtils.getApplicationIdFromJWT(verifiedJwt);
+        FhirContext fhirContext = FhirContext.forR4();
+        IGenericClient client = PermissionAccessChecker.createFhirClientForR4(fhirContext);
+
+        Bundle compositionBundle =
+                client.search()
+                        .forResource(Composition.class)
+                        .where(Composition.IDENTIFIER.exactly().identifier(applicationId))
+                        .returnBundle(Bundle.class)
+                        .execute();
+
+        Bundle.BundleEntryComponent compositionEntry = compositionBundle.getEntryFirstRep();
+        Composition composition;
+        if (compositionEntry != null) {
+            composition = (Composition) compositionEntry.getResource();
+        } else {
+            composition = null;
+        }
+        String binaryResourceReference =
+                PermissionAccessChecker.getBinaryResourceReference(composition);
+        Binary binary =
+                PermissionAccessChecker.readApplicationConfigBinaryResource(
+                        binaryResourceReference, fhirContext);
+        return PermissionAccessChecker.findSyncStrategy(binary);
+    }
```

Comment:
We can do some benchmarking here but I think there's an opportunity to optimise via caching by application ID. This will save us 2 API calls and the extra processing.